libshaderc/src/shaderc_test.cc

```diff
@@ -482,6 +483,190 @@ TEST_F(CompileStringWithOptionsTest, PreprocessingOnlyOption) {
+
+  IncluderResponsor::FakeFS fake_fs({
+      {"file_0",
+       {"path/to/file_0",
+        "#include \"file_1\"\n"
+        "void main() {foo(); bar();}\n"}},
+      {"file_1", {"path/to/file_1", "void bar() {}\n"}},
+  });
+
+  // Sets callbacks for libshaderc includer.
+  IncluderResponsor responsor(options_.get(), fake_fs);
+  // Clone a new options object.
+  compile_options_ptr cloned_options(
+      shaderc_compile_options_clone(options_.get()));
+  EXPECT_TRUE(CompilesToValidSpv(kMainIncludingShader,
+                                 shaderc_glsl_vertex_shader, cloned_options.get()));
+  // Sets compiler to preprocessing only mode, so we can check the result after
+  // preprocessing. Expect to see the content of the fake file object in the
+  // output of preprocessing.
+  shaderc_compile_options_set_preprocessing_only_mode(cloned_options.get());
+  std::string preprocessing_output = CompilationOutput(
+      kMainIncludingShader, shaderc_glsl_vertex_shader, cloned_options.get());
+  EXPECT_THAT(preprocessing_output, HasSubstr("#line 0 \"path/to/file_0\"\n"
+                                              "#line 0 \"path/to/file_1\"\n"
+                                              " void bar(){ }\n"
+                                              "#line 1 \"path/to/file_0\"\n"
+                                              " void main(){ foo();bar();}\n"
+                                              "#line 2"));
+}
+
+TEST_F(CompileStringWithOptionsTest, IncluderNestedIncludeClonedOptions) {
+  const std::string kMainIncludingShader =
+      "void foo() {}\n"
+      "#include \"file_0\"\n";
+
+  IncluderResponsor::FakeFS fake_fs({
+      {"file_0",
+       {"path/to/file_0",
+        "#include \"file_1\"\n"
+        "void main() {foo(); bar();}\n"}},
+      {"file_1", {"path/to/file_1", "void bar() {}\n"}},
+  });
+
+  // Sets callbacks for libshaderc includer.
+  IncluderResponsor responsor(options_.get(), fake_fs);
+
+  EXPECT_TRUE(CompilesToValidSpv(kMainIncludingShader,
+                                 shaderc_glsl_vertex_shader, options_.get()));
+  // Sets compiler to preprocessing only mode, so we can check the result after
+  // preprocessing. Expect to see the content of the fake file object in the
+  // output of preprocessing.
+  shaderc_compile_options_set_preprocessing_only_mode(options_.get());
+  std::string preprocessing_output = CompilationOutput(
+      kMainIncludingShader, shaderc_glsl_vertex_shader, options_.get());
+  EXPECT_THAT(preprocessing_output, HasSubstr("#line 0 \"path/to/file_0\"\n"
+                                              "#line 0 \"path/to/file_1\"\n"
+                                              " void bar(){ }\n"
+                                              "#line 1 \"path/to/file_0\"\n"
+                                              " void main(){ foo();bar();}\n"
+                                              "#line 2"));
+}
+
 TEST_F(CompileStringWithOptionsTest, WarningsOnLine) {
```

Comment:
Similarly, should implement the same suite of tests as the C++ API tests.  And use a single test implementation but parameterized by the fake filesystem contents, initial filename, and expected preprocessor output.